---
title: "Using qgg"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using qgg}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


## Installation

You can install **qgg** from github with:

```{r gh-installation, eval = FALSE}
# install.packages("devtools")
library(devtools)
options(devtools.install.args=" --no-multiargs")
devtools::install_github("psoerensen/qgg")
```

Windows users may encounter a problem during installation using devtools (error message: "Error in read.dcf(path) :"). 
Until the devtools issue is resolved you should be able to install using the following lines:
```{r, eval = FALSE}
library(remotes)
install_url(url="https://github.com/psoerensen/qgg/archive/master.zip", INSTALL_opt= "--no-multiarch")
```
<br>
  
## Small data set tutorial

This detailed tutorial on a small data set, exemplifies some of the core features of **qgg**. It takes you all the way from getting and cleaning data to estimation of genomic parameters, gene set enrichment analysis and making predictions with genomic best linear unbiased prediction (GBLUP) and genomic feature BLUP (GFBLUP) models.  
  
The phenotypic and genetic data is from the publicly available [*Drosophila* genetic reference panel](http://dgrp2.gnets.ncsu.edu/) (DGRP) - a system of 205 completely inbred *D. melanogaster* lines.  
  
The tutorial is divided into the following steps:
  
* **Data preparation**  
  
     + [Prepare phenotype and covariate data](http://psoerensen.github.io/qgg/articles/prepare_phenotype_covariate_data.html)  
     
     + [Prepare genotype data](http://psoerensen.github.io/qgg/articles/prepare_genotype_data.html)  
     
     + [Prepare annotation data and marker sets](http://psoerensen.github.io/qgg/articles/annotation_and_marker_sets.html)  
  
  
* **Data analysis**  

     + [Prediction using GBLUP and cross validation](http://psoerensen.github.io/qgg/articles/GBLUP_cross_validation.html)  
  
     + [Prediction using GFBLUP and cross validation](http://psoerensen.github.io/qgg/articles/GFBLUP_cross_validation.html)  
  
     + [Gene set enrichment analysis](http://psoerensen.github.io/qgg/articles/gsea.html)
     
## Large data set tutorials

These condensed tutorials on two large data sets, exemplify some of the core features of **qgg**. It takes you all the way from getting and cleaning data to estimation of genomic parameters, gene set enrichment analysis and making predictions with genomic best linear unbiased prediction (GBLUP) and genomic feature BLUP (GFBLUP) models.  
  
* [Tutorial on using 1000 Genomes Project data](http://psoerensen.github.io/qgg/articles/1000genome_tutorial.html) 

* [Prepare UK Biobank phenotype data](http://psoerensen.github.io/qgg/articles/prepare_ukb_phenotypes.html)

* [Tutorial on using UK Biobank data](http://psoerensen.github.io/qgg/articles/ukb_tutorial.html)  

     
